core/unix/coredump.c

```diff
@@ -154,6 +194,178 @@ write_section_header(DR_PARAM_IN file_t elf_file,
+    regs->regs[28] = mcontext->r28;
+    regs->regs[29] = mcontext->r29;
+    regs->regs[30] = mcontext->r30;
+    regs->sp = mcontext->sp;
+    regs->pc = (uint64_t)mcontext->pc;
+#endif
+}
+
+/*
+ * Write prstatus structure to the file. Returns true if the note is written to
+ * the file, false otherwise.
+ */
+static bool
+write_prstatus_note(DR_PARAM_IN file_t elf_file)
+{
+    void *drcontext = dr_get_current_drcontext();
+    dr_mcontext_t mcontext;
+    mcontext.size = sizeof(mcontext);
+    mcontext.flags = DR_MC_ALL;
+    dr_get_mcontext(drcontext, &mcontext);
+
+    struct elf_prstatus prstatus;
+    struct user_regs_struct *regs = (struct user_regs_struct *)&prstatus.pr_reg;
+    mcontext_to_user_regs(&mcontext, regs);
+
+    ELF_NOTE_HEADER_TYPE nhdr;
+    // Add one to include the terminating null character.
+    nhdr.n_namesz = strlen(NOTE_OWNER) + 1;
+    nhdr.n_descsz = sizeof(prstatus);
+    nhdr.n_type = NT_PRSTATUS;
+    if (os_write(elf_file, &nhdr, sizeof(nhdr)) != sizeof(nhdr)) {
+        return false;
+    }
+    if (os_write(elf_file, NOTE_OWNER, NOTE_OWNER_LENGTH) != NOTE_OWNER_LENGTH) {
+        return false;
+    }
+    return os_write(elf_file, &prstatus, sizeof(prstatus)) == sizeof(prstatus);
+}
+
+/*
+ * Copy register values from dr_mcontext_t to elf_fpregset_t for AARCH64. For
+ * X86, copy the floating point registers from the output of fxsave64.
+ */
+static bool
+get_floating_point_registers(DR_PARAM_IN dr_mcontext_t *mc,
+                             DR_PARAM_OUT elf_fpregset_t *regs)
+{
+#ifdef DR_HOST_NOT_TARGET
+    return false;
+#elif defined(X86)
+    byte fpstate_buf[MAX_FP_STATE_SIZE];
+    fxsave64_map_t *fpstate = (fxsave64_map_t *)ALIGN_FORWARD(fpstate_buf, 16);
+    if (proc_save_fpstate((byte *)fpstate) != DR_FPSTATE_BUF_SIZE) {
+        return false;
+    }
+
+    regs->cwd = fpstate->fcw;
+    regs->swd = fpstate->fsw;
+    regs->ftw = fpstate->ftw;
+    regs->fop = fpstate->fop;
+    regs->mxcsr = fpstate->mxcsr;
+    regs->mxcr_mask = fpstate->mxcsr;
+    /* 8*16 bytes for each FP-reg = 128 bytes */
+    for (int i = 0; i < 32; ++i) {
+        regs->st_space[i] = fpstate->st_space[i];
+    }
+    /* 16*16 bytes for each XMM-reg = 256 bytes */
+    for (int i = 0; i < 64; ++i) {
+        regs->xmm_space[i] = fpstate->xmm_space[i];
+    }
+    return true;
+#elif defined(AARCH64)
+    regs->fpsr = mc->fpsr;
+    regs->fpcr = mc->fpcr;
+    for (int i = 0; i < MCXT_NUM_SIMD_SVE_SLOTS; ++i) {
+        memcpy(&regs->vregs[i], &mc->simd[i].q, sizeof(mc->simd[i].q));
+    }
+    return true;
+#else
+    return false;
+#endif
+}
+
+/*
+ * Write floating point registers to the file. Returns true if the note is written to
+ * the file, false otherwise.
+ */
+static bool
+write_fpregset_note(DR_PARAM_IN file_t elf_file)
+{
+    void *drcontext = dr_get_current_drcontext();
+    dr_mcontext_t mcontext;
+    mcontext.size = sizeof(mcontext);
+    mcontext.flags = DR_MC_ALL;
```

Comment:
You don't need integer regs, right?

Reply:
No, integer registers are not needed here. I've updated to code to call dr_get_mcontext_priv() from dr_create_memory_dump() instead.